i04-Scanner

Rujie92 opened the pull request:
Contains an example using Bresenham algorithm to draw a line when scanning

Rujie92 committed:
Change map initialization

Rujie92 committed:
Delete Robot.Motion.mat

Rujie92 committed:
Delete scanner.m

Rujie92 committed:
Add scanner that uses bresenham algorithm

Rujie92 committed:
A function to detect robot collision

Rujie92 committed:
A function to create a map

Rujie92 committed:
A function to generate robot motions

Rujie92 committed:
An example of scanner using Bresenham algorithm

Rujie92 committed:
Rename this file to be a ".m" file

Rujie92 committed:
Rename this file to be a ".m" file

Rujie92 committed:
Rename this file to be a ".m" file

Rujie92 committed:
Rename this file to be a ".m" file

Rujie92 committed:
Rename this file to be a ".m" file

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -0,0 +1,101 @@
+     title('Real map & Robot positions');
+     
+   if (makemovie1) 
+       writeVideo(vidObj1, getframe(gca)); end
```
I ran the example and checked the movies, they are identical. I think lines 92 & 93 should be moved up to around line 83 depending on what was supposed to be captured.

Rujie92 replied:
Yeah,  I just want to show the robot position in an explored map for reviewers to compare.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -0,0 +1,101 @@
+     colormap('gray');
+     plot(X(2,i),X(1,i),'x','Markersize',10);
+     plot(X(2,1:i),X(1,1:i),'-','Linewidth',2);
+     title('Real map & Robot positions');
```
The two figures do not share the same axis which makes the videos misaligned. I suggest copying your great line of code from line 77 and putting it here too :)

Rujie92 replied:
Thank you, I didn't notice this @ _ @

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/create_a_map.m:
```diff
@@ -0,0 +1,24 @@
+function [ map ] = create_a_map( M, N )
```
Great function! maybe it should be placed in "matlab_simulation/10-environments" for use with other examples. The issue I worked on could certainly make good use of it.

Rujie92 replied:
Sure, I will move it later.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -0,0 +1,101 @@
+     inv_mm = exp(inv_mm) ./ (1+exp(inv_mm));
+         
+   % Plot Inverse Measurement Model
+     figure(1);clf;hold on;
```
Minor thing, but it looks like something weird happened to the formatting here. The code is indented compared to the comments, not a big deal, but it is inconsistent with earlier in the document.

Rujie92 replied:
Actually, I found a error in the codes and this error will affect here. Will deal with it later.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/robot_motion.m:
```diff
@@ -0,0 +1,108 @@
+ X(:,1) = x0; % Initial state of the robot
+ 
+ %%
+    mode = 5; % Robot initial moving mode
```
Something happened with the spacing again here. Could use a bit of cleaning up, otherwise great.

Rujie92 replied:
Same as the last problem, will do a lot of cleaning up.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/scanner_bres.m:
```diff
@@ -0,0 +1,75 @@
+clear;clc;
```
I'm not sure why this file now exists. You example.m file appears to do the same thing and does a pretty good job. Perhaps this can be removed.

Rujie92 replied:
Yes, no need to have this one.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -0,0 +1,101 @@
+   
+   for j = 1 : length(meas_r) 
+      % Compute inverse measurement model
+      inv_meas_mod = inversescannerbres(M, N, X(1,i), X(2,i), X(3,i) + meas_phi(j), meas_r(j), rmax);
```
Nothing wrong here but I thought I would mention: I also used the existing function "inversescannerbres" for the occupancy grid mapping issue I worked on. I happened to find a bug in it on line 16 where the "length(list)" would create a matrix of the wrong size if the range measurements were really tiny (up against a wall or something). Anyway, I made an updated version of the function called "inverse_scanner_bres.m", you may want to consider it, yay code re-use!

Rujie92 replied:
Thanks, I'll consider to modify this place and try the situation of a tiny measurement range.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/robot_motion.m:
```diff
@@ -0,0 +1,108 @@
+    y_vel = 2*round(0.4*randn(1));
+ end
+
+ [M N] = size(map);
```
MatLab toolips are telling me that M and N are unused as well as mode (line 25), consider removing them.

Rujie92 replied:
Yes, they are some residuals after debuging that day.

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/collision_pred.m:
```diff
@@ -0,0 +1,74 @@
+    if ((pred_x == 1) || (pred_x == N)) % If the robot hit the vertical boundaries
+        if x_vel<0
+          coordinate = [ 1, map( pred_x,pred_y+1), map( pred_x+1, pred_y), map( pred_x, pred_y-1) ]; % Left boundary
+        else x_vel>0
```
Should this line and line 30 be elseif rather than else? The code seemed to work which is weird.....

Rujie92 replied:
Thanks, What I thought is to judge the way the robot hit the wall. Form "Line 19", if it hits the vertical boundaries, it could only have two situations which are hitting towards left ( x_vel < 0 ) and hitting towards right ( x_vel > 0 ). So an else would be enough I thought?

icolwell commented on matlab_simulation/04-sensor/scanner_bresenham/robot_motion.m:
```diff
@@ -0,0 +1,108 @@
+function [X] = robot_motion( map, X, x0, T)
```
I'm wondering if this function and "collision_pred" should be combined? I didn't dig in too much but it looks like they do a lot of similar logic. Just a thought.

Rujie92 replied:
Great idea,I could try to merge them but a huge code always make me dizzy...

Chiasung commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -0,0 +1,101 @@
+   end
+   
+   % Transform inverse measurement model from log-odds form to pobability
+     inv_mm = exp(inv_mm) ./ (1+exp(inv_mm));
```
"Inverse model" seems not good variable name to add to the equation. It's better to use clear and different variable names on both side of this equation. Such like left hand side is P (probability) and logitP (Log odds)is in the exp function.

Rujie92 replied:
Yeah, good idea for clear coding.

Chiasung commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -0,0 +1,101 @@
+   % Plot real map and robot position
+     figure(2);clf;hold on;
+     image(100 * (1-map_real));
+     colormap('gray');
```
Overall this is a  good example. However, the colour of  the laser scan lines are white.  It's better to use the black line to represent the ray as we see on the slides.

Rujie92 replied:
Yeah, a big error inside the code. Debug it later.

Rujie92 committed:
Moving creat_a_map into 10-environment folder

Rujie92 committed:
Move this function file

This function file has been placed into folder 10-environment so that it can be used by other examples.

Rujie92 committed:
Update collision_pred.m

Did some cleaning up and corrected typos

Rujie92 committed:
Update example.m

-Added a new firgure showing the occupancy grid mapping
-Did some cleanning up jobs
-Replaced inversescannerbres.m with inverse_scanner_bres.m modified by Mr. icolwell
-Renamed some variable names

Rujie92 committed:
Update robot_motion.m

- Did some cleanning up

Rujie92 committed:
Delete Scanner_bres.m

No need to have this file cuz it has the same content as the example.m.

Chiasung commented on matlab_simulation/04-sensor/scanner_bresenham/example.m:
```diff
@@ -25,10 +33,11 @@
 % Generate the belief map
```
## Just did a quickly check again. I got following error. 

Undefined function or variable 'create_a_map'.

Error in example (line 31)
map_real = create_a_map(M,N); % Create a random map with obstacles

Rujie92 replied:
That "creat_a_map.m" function is in the 10-environment folder, I put in there so that everyone could use it as a part of their codes.

Rujie92 committed:
Update example.m

Correct some typos

Rujie92 committed:
Create readme.txt

An explanation of the codes
